# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFontUtils.h
# Commit: 3b719cd12933
# Full Hash: 3b719cd1293370d843691b3656b52b04042d5ec4
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2022-03-02 04:31:47
# Regressor Bug: 1756720
# File Overlap Count: 1
# Description:
#   Bug 1756720 - When using legacy MS Symbol fonts, map character codes from the U+00xx range to the PUA range U+F0xx found in the cmap subtables. r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D139724
# ==============================================================================

diff -r 9849c6e79fa3 -r 3b719cd12933 gfx/thebes/gfxFontUtils.h
--- a/gfx/thebes/gfxFontUtils.h	Tue Mar 01 17:27:27 2022 +0200
+++ b/gfx/thebes/gfxFontUtils.h	Tue Mar 01 15:22:56 2022 +0000
@@ -980,7 +980,8 @@
                                             gfxSparseBitSet& aCharacterMap);
 
   static nsresult ReadCMAPTableFormat4(const uint8_t* aBuf, uint32_t aLength,
-                                       gfxSparseBitSet& aCharacterMap);
+                                       gfxSparseBitSet& aCharacterMap,
+                                       bool aIsSymbolFont);
 
   static nsresult ReadCMAPTableFormat14(const uint8_t* aBuf, uint32_t aLength,
                                         mozilla::UniquePtr<uint8_t[]>& aTable);
@@ -988,7 +989,8 @@
   static uint32_t FindPreferredSubtable(const uint8_t* aBuf,
                                         uint32_t aBufLength,
                                         uint32_t* aTableOffset,
-                                        uint32_t* aUVSTableOffset);
+                                        uint32_t* aUVSTableOffset,
+                                        bool* aIsSymbolFont);
 
   static nsresult ReadCMAP(const uint8_t* aBuf, uint32_t aBufLength,
                            gfxSparseBitSet& aCharacterMap,
@@ -1015,6 +1017,12 @@
   static uint32_t MapCharToGlyph(const uint8_t* aCmapBuf, uint32_t aBufLength,
                                  uint32_t aUnicode, uint32_t aVarSelector = 0);
 
+  // For legacy MS Symbol fonts, we try mapping 8-bit character codes to the
+  // Private Use range at U+F0xx used by the cmaps in these fonts.
+  static MOZ_ALWAYS_INLINE uint32_t MapLegacySymbolFontCharToPUA(uint32_t aCh) {
+    return aCh >= 0x20 && aCh <= 0xff ? 0xf000 + aCh : 0;
+  }
+
 #ifdef XP_WIN
   // determine whether a font (which has already been sanitized, so is known
   // to be a valid sfnt) is CFF format rather than TrueType